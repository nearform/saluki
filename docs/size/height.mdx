import { Height } from './'

# Height

---

## Example

<Height />

```js
const Box = styled.div`
  ${size.height.medium}
  ${background.color.black}
`
```

## Rules

```js
size.height.small
size.height.medium
size.height.large
size.height.quarter
size.height.half
size.height.threeQuarters
size.height.full
size.height.view
size.height.auto
```

## Configuration

Configurable via;

```
size.height
```

## Defaults

```js
{
  small: '2rem',
  medium: '4rem',
  large: '8rem',
  quarter: '25%',
  half: '50%',
  threeQuarters: '75%',
  full: '100%',
  view: '100vw',
  auto: 'auto'
}
```
