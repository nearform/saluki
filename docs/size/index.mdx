# Size

---

## Rules

```js
size.width.fifth
size.width.twoFifths
size.width.threeFifths
size.width.fourFifths
size.width.quarter
size.width.threeQuarters
size.width.third
size.width.twoThirds
size.width.half
size.width.full
size.width.view
size.width.auto

size.height.small
size.height.medium
size.height.large
size.height.quarter
size.height.half
size.height.threeQuarters
size.height.full
size.height.view
size.height.auto
```

## Configuration

Configurable via;

```
size
```

## Defaults

```js
{
  width: {
    fifth: '20%',
    twoFifths: '40%',
    threeFifths: '60%',
    fourFifths: '80%',
    quarter: '25%',
    threeQuarters: '75%',
    third: '33.3333333333%',
    twoThirds: '66.6666666666%',
    half: '50%',
    full: '100%',
    view: '100vw',
    auto: 'auto'
  },
  height: {
    small: '2rem',
    medium: '4rem',
    large: '8rem',
    quarter: '25%',
    half: '50%',
    threeQuarters: '75%',
    full: '100%',
    view: '100vw',
    auto: 'auto'
  }
}
```
