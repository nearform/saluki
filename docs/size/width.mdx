import { Width } from './'

# Width

---

## Example

<Width>Hello</Width>

```js
const Box = styled.div`
  ${size.width.half}
  ${background.color.red}
`
```

## Rules

```js
size.width.fifth
size.width.twoFifths
size.width.threeFifths
size.width.fourFifths
size.width.quarter
size.width.threeQuarters
size.width.third
size.width.twoThirds
size.width.half
size.width.full
size.width.view
size.width.auto
```

## Configuration

Configurable via;

```
size.width
```

## Defaults

```js
{
  fifth: '20%',
  twoFifths: '40%',
  threeFifths: '60%',
  fourFifths: '80%',
  quarter: '25%',
  threeQuarters: '75%',
  third: '33.3333333333%',
  twoThirds: '66.6666666666%',
  half: '50%',
  full: '100%',
  view: '100vw',
  auto: 'auto'
}
```
