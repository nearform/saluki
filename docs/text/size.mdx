import { Size } from './'

# Text Size

---

## Example

<Size>Hello</Size>

```js
const Text = styled.h1`
  ${text.size.large}
`
```

## Configuration

Configurable via;

```
text.size
```
