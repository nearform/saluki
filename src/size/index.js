export const defaultRules = {
  width: {
    fifth: '20%',
    twoFifths: '40%',
    threeFifths: '60%',
    fourFifths: '80%',
    quarter: '25%',
    threeQuarters: '75%',
    third: '33.3333333333%',
    twoThirds: '66.6666666666%',
    half: '50%',
    full: '100%',
    view: '100vw',
    auto: 'auto'
  },
  height: {
    small: '2rem',
    medium: '4rem',
    large: '8rem',
    quarter: '25%',
    half: '50%',
    threeQuarters: '75%',
    full: '100%',
    view: '100vw',
    auto: 'auto'
  }
}

export function combineRules(
  defaultRules,
  customRules = {
    width: {},
    height: {}
  }
) {
  return {
    width: {
      ...defaultRules.width,
      ...customRules.width
    },
    height: {
      ...defaultRules.height,
      ...customRules.height
    }
  }
}

function generateRules(rules, cssProp) {
  const newRules = {}

  for (const [key, value] of Object.entries(rules)) {
    newRules[key] = {
      [cssProp]: value
    }
  }

  return newRules
}

export default function(rules) {
  return {
    width: generateRules(rules.width, 'width'),
    height: generateRules(rules.height, 'height')
  }
}
