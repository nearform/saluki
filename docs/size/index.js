import styled from 'styled-components/macro'
import { size, background } from '../../src'

export const Width = styled.div`
  ${size.width.half}
  ${background.color.red}
`

export const Height = styled.div`
  ${size.height.medium}
  ${background.color.black}
`
