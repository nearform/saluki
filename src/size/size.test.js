import size, { defaultRules, combineRules } from './'

describe('size', () => {
  describe('when no custom config is provided', () => {
    it('should return the default config', () => {
      const combinedRules = combineRules(defaultRules)
      expect(size(combinedRules)).toMatchSnapshot()
    })
  })

  describe('when custom config is provided', () => {
    describe('when the custom config overrides a value', () => {
      it('should replace the default value', () => {
        const combinedRules = combineRules(defaultRules, {
          height: { small: '10rem' }
        })
        expect(size(combinedRules)).toMatchSnapshot()
      })
    })

    describe('when the custom config provides a new value', () => {
      it('should add the value', () => {
        const combinedRules = combineRules(defaultRules, {
          width: { tenth: '10%' }
        })
        expect(size(combinedRules)).toMatchSnapshot()
      })
    })
  })
})
