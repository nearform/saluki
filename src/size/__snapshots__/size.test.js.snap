// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`size when custom config is provided when the custom config overrides a value should replace the default value 1`] = `
Object {
  "height": Object {
    "auto": Object {
      "height": "auto",
    },
    "full": Object {
      "height": "100%",
    },
    "half": Object {
      "height": "50%",
    },
    "large": Object {
      "height": "8rem",
    },
    "medium": Object {
      "height": "4rem",
    },
    "quarter": Object {
      "height": "25%",
    },
    "small": Object {
      "height": "10rem",
    },
    "threeQuarters": Object {
      "height": "75%",
    },
    "view": Object {
      "height": "100vw",
    },
  },
  "width": Object {
    "auto": Object {
      "width": "auto",
    },
    "fifth": Object {
      "width": "20%",
    },
    "fourFifths": Object {
      "width": "80%",
    },
    "full": Object {
      "width": "100%",
    },
    "half": Object {
      "width": "50%",
    },
    "quarter": Object {
      "width": "25%",
    },
    "third": Object {
      "width": "33.3333333333%",
    },
    "threeFifths": Object {
      "width": "60%",
    },
    "threeQuarters": Object {
      "width": "75%",
    },
    "twoFifths": Object {
      "width": "40%",
    },
    "twoThirds": Object {
      "width": "66.6666666666%",
    },
    "view": Object {
      "width": "100vw",
    },
  },
}
`;

exports[`size when custom config is provided when the custom config provides a new value should add the value 1`] = `
Object {
  "height": Object {
    "auto": Object {
      "height": "auto",
    },
    "full": Object {
      "height": "100%",
    },
    "half": Object {
      "height": "50%",
    },
    "large": Object {
      "height": "8rem",
    },
    "medium": Object {
      "height": "4rem",
    },
    "quarter": Object {
      "height": "25%",
    },
    "small": Object {
      "height": "2rem",
    },
    "threeQuarters": Object {
      "height": "75%",
    },
    "view": Object {
      "height": "100vw",
    },
  },
  "width": Object {
    "auto": Object {
      "width": "auto",
    },
    "fifth": Object {
      "width": "20%",
    },
    "fourFifths": Object {
      "width": "80%",
    },
    "full": Object {
      "width": "100%",
    },
    "half": Object {
      "width": "50%",
    },
    "quarter": Object {
      "width": "25%",
    },
    "tenth": Object {
      "width": "10%",
    },
    "third": Object {
      "width": "33.3333333333%",
    },
    "threeFifths": Object {
      "width": "60%",
    },
    "threeQuarters": Object {
      "width": "75%",
    },
    "twoFifths": Object {
      "width": "40%",
    },
    "twoThirds": Object {
      "width": "66.6666666666%",
    },
    "view": Object {
      "width": "100vw",
    },
  },
}
`;

exports[`size when no custom config is provided should return the default config 1`] = `
Object {
  "height": Object {
    "auto": Object {
      "height": "auto",
    },
    "full": Object {
      "height": "100%",
    },
    "half": Object {
      "height": "50%",
    },
    "large": Object {
      "height": "8rem",
    },
    "medium": Object {
      "height": "4rem",
    },
    "quarter": Object {
      "height": "25%",
    },
    "small": Object {
      "height": "2rem",
    },
    "threeQuarters": Object {
      "height": "75%",
    },
    "view": Object {
      "height": "100vw",
    },
  },
  "width": Object {
    "auto": Object {
      "width": "auto",
    },
    "fifth": Object {
      "width": "20%",
    },
    "fourFifths": Object {
      "width": "80%",
    },
    "full": Object {
      "width": "100%",
    },
    "half": Object {
      "width": "50%",
    },
    "quarter": Object {
      "width": "25%",
    },
    "third": Object {
      "width": "33.3333333333%",
    },
    "threeFifths": Object {
      "width": "60%",
    },
    "threeQuarters": Object {
      "width": "75%",
    },
    "twoFifths": Object {
      "width": "40%",
    },
    "twoThirds": Object {
      "width": "66.6666666666%",
    },
    "view": Object {
      "width": "100vw",
    },
  },
}
`;
